File: src/main/java/gg/mineral/practice/inventory/menus/SelectCategorizedGametypeMenu.java
package gg.mineral.practice.inventory.menus;

import gg.mineral.practice.category.Category;
import gg.mineral.practice.entity.PlayerStatus;
import gg.mineral.practice.gametype.Gametype;
import gg.mineral.practice.inventory.ClickCancelled;
import gg.mineral.practice.queue.Queuetype;
import gg.mineral.practice.queue.QueuetypeMenuEntry;
import gg.mineral.practice.util.messages.CC;
import it.unimi.dsi.fastutil.objects.Object2IntLinkedOpenHashMap;

@ClickCancelled(true)
public class SelectCategorizedGametypeMenu extends SelectGametypeMenu {
    protected final Category category;

    public SelectCategorizedGametypeMenu(Queuetype queuetype, Category category, Type type) {
        super();
        this.category = category;
        this.queuetype = queuetype;
        this.type = type;
        this.menuEntries = setMenuEntries();
    }

    @Override
    protected Object2IntLinkedOpenHashMap<QueuetypeMenuEntry> setMenuEntries() {
        Object2IntLinkedOpenHashMap<QueuetypeMenuEntry> menuEntries = new Object2IntLinkedOpenHashMap<>();
        category.getGametypes().forEach(gametype -> {
            if (gametype.isInCategory())
                menuEntries.put(gametype, queuetype.getMenuEntries().getInt(gametype));
        });
        return menuEntries;
    }

    @Override
    public String getTitle() {
        return CC.BLUE + category.getName();
    }

    @Override
    protected boolean shouldSkip(QueuetypeMenuEntry menuEntry) {
        return menuEntry instanceof Gametype gametype && !gametype.isInCategory();
    }

    @Override
    public void onClose() {
        if (viewer.getPlayerStatus() == PlayerStatus.FIGHTING || viewer.getOpenMenu() != null)
            return;

        viewer.openMenu(new SelectGametypeMenu(queuetype, type));
    }
}

File: src/main/kotlin/gg/mineral/practice/inventory/menus/SelectCategorizedGametypeMenu.kt
package gg.mineral.practice.inventory.menus

import gg.mineral.practice.category.Category
import gg.mineral.practice.entity.PlayerStatus
import gg.mineral.practice.gametype.Gametype
import gg.mineral.practice.inventory.ClickCancelled
import gg.mineral.practice.inventory.Menu
import gg.mineral.practice.managers.GametypeManager
import gg.mineral.practice.queue.Queuetype
import gg.mineral.practice.queue.QueuetypeMenuEntry
import gg.mineral.practice.util.messages.CC
import it.unimi.dsi.fastutil.objects.Object2IntLinkedOpenHashMap

@ClickCancelled(true)
class SelectCategorizedGametypeMenu(queuetype: Queuetype, val category: Category, type: Type, prevMenu: Menu? = null) :
    SelectGametypeMenu(queuetype, type, prevMenu) {

    override fun setMenuEntries(): Object2IntLinkedOpenHashMap<QueuetypeMenuEntry> {
        val menuEntries = Object2IntLinkedOpenHashMap<QueuetypeMenuEntry>()
        category.gametypes.mapNotNull { GametypeManager.gametypes[it] }.forEach { gametype ->
            if (gametype.inCategory) menuEntries.put(
                gametype,
                queuetype.menuEntries.getInt(gametype)
            )
        }
        return menuEntries
    }

    override val title: String
        get() = CC.BLUE + category.name

    override fun shouldSkip(menuEntry: QueuetypeMenuEntry) = menuEntry is Gametype && !menuEntry.inCategory

    override fun onClose() {
        if (viewer.playerStatus === PlayerStatus.FIGHTING || viewer.playerStatus === PlayerStatus.KIT_CREATOR || viewer.playerStatus === PlayerStatus.KIT_EDITOR || viewer.openMenu != null) return
        viewer.openMenu(SelectGametypeMenu(queuetype, type, null))
    }
}
